### PR TITLE
Remove the complex IAccurateTagger codepaths from the generic async tagger implementation.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We have one use case where we want to support IAccurateTagger, but currently all async taggers are both automatically opted into accurate tagging *and* need all the internal complexity for it.  That case is for outlining, which wants to be able to get tags accurately so it can collapse spans on open.  However, by default the editor keeps the spans collapsed that were already collapsed before. So the only places we need to precompute the spans that shoudl be collapsed (because the editor has no persisted state for them) are:

1. Metadata-as-source.  These files are always fresh, so we have to compute the spans we want to collapse to up front.
2. A file with a `#region` in it.  If this is the first time we're opening such a file we want to collapse those sections (but from taht point on, the editor will determine behavior based on what's in the SUO file).

Because this is entirely related to outlining, we can move this customization solely into that scenario, and solely for the case of getting *the initial* tags accurately, not any tags beyond that.  Once we've computed those tags, we can switch entirely over to async tagging.

This helps out the IDE as the editor will defer to IAccurateTagger if it is exposed for outlining if that's what is there.  We do not want all outlining calls on file open to pay the price of forcing us to parse and whatnot.  So we change it so by default we do not pay, and we only have these costs for the special cases.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/Structure/AbstractStructureTaggerProvider.cs [hunk 1]
```diff
@@ -50,6 +53,50 @@ internal abstract partial class AbstractStructureTaggerProvider :
             ProjectionBufferFactoryService = projectionBufferFactoryService;
         }
 
+        protected override bool ComputeInitialTagsSynchronously(ITextBuffer subjectBuffer)
+        {
+            // If we can't find this doc, or outlining is not enabled for it, no need to computed anything synchronously.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Structure/AbstractStructureTaggerProvider.cs:58]: this logic moved from AbstractLanguageService`2

File: src/EditorFeatures/Core/Implementation/Structure/AbstractStructureTaggerProvider.cs [hunk 2]
```diff
@@ -50,6 +53,50 @@ internal abstract partial class AbstractStructureTaggerProvider :
...
+                }
+
+                return false;
+
+                static bool StartsWithRegionTag(ITextSnapshotLine line)
+                {
+                    var start = line.GetFirstNonWhitespacePosition();
+                    return start != null && line.StartsWith(start.Value, "#region", ignoreCase: true);
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Structure/AbstractStructureTaggerProvider.cs:97]: not great that we have this linear scan on all outlining taggers.  but that matches exactly what we've been doing so far.  so we're no worse off :(

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 3]
```diff
@@ -193,21 +184,6 @@ private object State
...
-                    _workQueue.AssertIsForeground();
-                    return _upToDate_doNotAccessDirectly;
-                }
-
-                set
-                {
-                    _workQueue.AssertIsForeground();
-                    _upToDate_doNotAccessDirectly = value;
-                }
-            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs:209]: the previous logic was more complex and required us to keep track if there were outstanding operations in flight that we had to cancel and refire.  We also didn't want to recompute if our comptuation was done and no more events had been heard.  But now, we're not implementing IAccurateTagger anymore, so we don't need to provide those semantics.  It's fine to return stale results all the time *except* for the two cases in outlinign we need it.

**Comment by CyrusNajmabadi** [comment]
@ryzngard @genlu can you ptal?

**Comment by CyrusNajmabadi** [comment]
@davidwengier if convenient :)

**Review by davidwengier — APPROVED** [review]

**Pull request merged**